File: fibonacci/csharp/fibonacci/Fibonacci.cs
using System;
using System.Collections.Immutable;

//0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811
namespace fibonacci
{
    internal class Fibonacci
    {
        public static void Main(string[] args)
        {
            // Imperative, mutable, and side-effecting
            void Fibs1(int len)
            {
                int a = 0;
                int b = 1;
                int i = 0;
                while (i < len)
                {
                    // Can we reorder any of these statements?
                    // Can we inline any of them?
                    // What refactorings are safe?
                    Console.Write(a);
                    if (i + 1 < len) Console.Write(", ");
                    int c = a + b;
                    a = b;
                    b = c;
                    i++;
                }
            }

            Fibs1(20);

            Console.WriteLine();

            // Separate out the side effects
            int[] Fibs2(int len)
            {
                int a = 0;
                int b = 1;
                int[] array = new int[len];
                int i = 0;
                while (i < len)
                {
                    array[i] = a;
                    int c = a + b;
                    a = b;
                    b = c;
                    i++;
                }
                return array;
            }

            Console.WriteLine(String.Join(", ", Fibs2(20)));

            // Move mutability to the top: variables are not themselves mutable
            ImmutableList<int> Fib3(int len)
            {
                int a = 0;
                int b = 1;
                ImmutableList<int> list = ImmutableList<int>.Empty;
                int i = 0;
                while (i < len)
                {
                    list = list.Add(a);
                    int c = a + b;
                    a = b;
                    b = c;
                    i++;
                }
                return list;
            }

            Console.WriteLine(String.Join(", ", Fib3(20)));

            // Convert to functional
            // 1. State we need to read inside the loop becomes a parameter
            // 2. State we need to read after the loop becomes the return value
            // 3. The looping condition determines whether to continue recursing
            // 4. Initial state becomes arguments to outer call, or argument defaults
            ImmutableList<int> Fib4(int len)
            {
                ImmutableList<int> Loop(ImmutableList<int> list, int i = 0, int a = 0, int b = 1)
                {
                    if (i < len)
                        return Loop(list.Add(a), i + 1, b, a + b);
                    else
                        return list;
                }

                return Loop(ImmutableList<int>.Empty);
            }

            Console.WriteLine(String.Join(", ", Fib4(20)));
        }
    }
}

File: fibonacci/scala/src/main/scala/fibonacci/Fibonacci.scala
package fibonacci

object Fibonacci {
  def main(args: Array[String]): Unit = {
    // Imperative, mutable, and side-effecting
    def fibs1(len: Int): Unit = {
      var a = 0
      var b = 1
      var i = 0
      while (i < len) {
        // Can we reorder any of these statements?
        // Can we inline any of them?
        // What refactorings are safe?
        print(a)
        if (i + 1 < len) print(", ")
        val c = a + b
        a = b
        b = c
        i += 1
      }
    }

    fibs1(20)

    println()

    // Separate out the side effects
    def fibs2(len: Int) = {
      var a = 0
      var b = 1
      val array = new Array[Int](len)
      var i = 0
      while (i < len) {
        array(i) = a
        val c = a + b
        a = b
        b = c
        i += 1
      }
      array
    }

    println(fibs2(20).mkString(", "))

    // Move mutability to the top: variables are not themselves mutable
    def fibs3(len: Int) = {
      var a = 0
      var b = 1
      var list = Seq.empty[Int]
      var i = 0
      while (i < len) {
        list = list :+ a
        val c = a + b
        a = b
        b = c
        i += 1
      }
      list
    }

    println(fibs3(20).mkString(", "))

    // Convert to functional
    // 1. State we need to read inside the loop becomes a parameter
    // 2. State we need to read after the loop becomes the return value
    // 3. The looping condition determines whether to continue recursing
    // 4. Initial state becomes arguments to outer call, or argument defaults
    def fib4(len: Int) = {
      def loop(list: Seq[Int] = Nil, i: Int = 0, a: Int = 0, b: Int = 1): Seq[Int] =
        if (i < len)
          loop(list :+ a, i + 1, b, a + b)
        else
          list

      loop()
    }

    println(fib4(20).mkString(", "))
  }
}
